In [1]:
import numpy as np
import pandas as pd

from sklearn.metrics import mean_absolute_error, mean_squared_error, root_mean_squared_error, r2_score

from sklearn.dummy import DummyRegressor

# Regression Metrics Example with Ice Cream Sales Data

## Example Data

| Temperature (°C) | Actual Ice Cream Sales | Predicted Ice Cream Sales |
|------------------|------------------------|---------------------------|
| 20               | 100                    | 110                       |
| 22               | 150                    | 140                       |
| 25               | 200                    | 210                       |
| 30               | 300                    | 290                       |
| 35               | 400                    | 410                       |



In [2]:
temperature = np.array([20, 22, 25, 30, 35])
actual_sales = np.array([100, 150, 200, 300, 400])
predicted_sales = np.array([110, 140, 210, 290, 410])

df = pd.DataFrame({
    'temperature':temperature,
    'actual_sales':actual_sales,
    'predicted_sales':predicted_sales,
})
df


temperature  actual_sales  predicted_sales
0           20           100              110
1           22           150              140
2           25           200              210
3           30           300              290
4           35           400              410

## Mean Absolute Error (MAE)

MAE is the average of the absolute differences between actual and predicted values. It measures the average magnitude of the errors in a set of predictions, without considering their direction.

**Formula:**
$$ \text{MAE} = \frac{1}{n} \sum_{i=1}^n |y_i - \hat{y}_i| $$

**Calculation:**

$$
\text{MAE} = \frac{|100 - 110| + |150 - 140| + |200 - 210| + |300 - 290| + |400 - 410|}{5} = \frac{10 + 10 + 10 + 10 + 10}{5} = 10
$$

In [3]:
# let's check what sklearn calculation is:
mae = mean_absolute_error(y_true=df['actual_sales'], y_pred=df['predicted_sales'])
mae

10.0

## Mean Squared Error (MSE)

MSE is the average of the squares of the errors. It measures the average squared difference between the actual and predicted values. Squaring the errors gives more weight to larger errors.

**Formula:**
$$ \text{MSE} = \frac{1}{n} \sum_{i=1}^n (y_i - \hat{y}_i)^2 $$

**Calculation:**

$$
\text{MSE} = \frac{(100 - 110)^2 + (150 - 140)^2 + (200 - 210)^2 + (300 - 290)^2 + (400 - 410)^2}{5} = \frac{100 + 100 + 100 + 100 + 100}{5} = 100
$$

In [4]:
# let's check what sklearn calculation is:
mse = mean_squared_error(y_true=df['actual_sales'], y_pred=df['predicted_sales'])
mse

100.0

## Root Mean Squared Error (RMSE)

RMSE is the square root of the average of the squared errors. It measures the square root of the average squared differences between actual and predicted values. RMSE is interpretable in the same units as the original data, making it more intuitive.

**Formula:**
$$ \text{RMSE} = \sqrt{\text{MSE}} $$

**Calculation:**

$$
\text{RMSE} = \sqrt{100} = 10
$$

In [5]:
# let's check what sklearn calculation is:
rmse = root_mean_squared_error(y_true=df['actual_sales'], y_pred=df['predicted_sales'])
rmse

10.0

## R-squared (Coefficient of Determination)
R-squared is a statistical measure that represents the proportion of the variance in the dependent variable that is predictable from the independent variables.

**Formula:**
$$
R^2 = 1 - \frac{\sum_{i=1}^n (y_i - \hat{y}_i)^2}{\sum_{i=1}^n (y_i - \bar{y})^2}
$$

**Calculation:**

1. Calculate the mean of the actual sales:
$$
\bar{y} = \frac{100 + 150 + 200 + 300 + 400}{5} = 230
$$

2. Calculate the total sum of squares (TSS):
$$
\sum_{i=1}^n (y_i - \bar{y})^2 = (100 - 230)^2 + (150 - 230)^2 + (200 - 230)^2 + (300 - 230)^2 + (400 - 230)^2 = 16900 + 6400 + 900 + 4900 + 28900 = 58000
$$

3. Calculate the residual sum of squares (RSS):
$$
\sum_{i=1}^n (y_i - \hat{y}_i)^2 = (100 - 110)^2 + (150 - 140)^2 + (200 - 210)^2 + (300 - 290)^2 + (400 - 410)^2 = 100 + 100 + 100 + 100 + 100 = 500
$$

4. Calculate R-squared:
$$
R^2 = 1 - \frac{500}{58000} \approx 0.9914
$$


In [6]:
# let's check what sklearn calculation is:
r2 = r2_score(y_true=df['actual_sales'], y_pred=df['predicted_sales'])
round(r2,4)

0.9914

### Interpretation
- **MAE**: The average absolute error is 10 ice creams. This means, on average, our predictions are off by 10 ice creams.
- **MSE**: The average of the squared errors is 100. Since the errors are squared, this value can be harder to interpret directly compared to MAE and RMSE.
- **RMSE**: The average error, considering the original scale of the data, is 10 ice creams. This provides a more intuitive measure of prediction error, similar to MAE but with more weight given to larger errors.
- **R-squared**: An R-squared value of approximately 0.9914 means that about 99.14% of the variability in ice cream sales can be explained by the model. The remaining 0.86% of the variability is due to other factors not included in the model.

## Evaluate the Errors

To determine whether an error, let's say Mean Absolute Error (MAE) is considered big or small, you need to consider the following factors:

1. Context of the Problem:

    Understand the domain and the specific problem you are addressing. Different domains have different tolerances for error.

2. Scale of the Target Variable:

    Compare the MAE to the range of the target variable. For instance, if the target variable ranges from 0 to 1000, an MAE of 2.34 might be quite small. However, if the target variable ranges from 0 to 10, an MAE of 2.34 is relatively large.

3. Benchmarking:

    Compare the MAE to benchmarks or previously reported results in similar studies or problems.

4. Baseline Model:

    Compare the MAE to a simple baseline model (e.g., mean predictor model). If the regression model significantly outperforms the baseline, the MAE might be acceptable.

5. Domain-Specific Acceptability:

    Consider the acceptable error margin in your specific domain. Some applications may require very low errors (e.g., medical predictions), while others may tolerate higher errors (e.g., movie recommendations)


### Steps to Evaluate the error (for example - MAE):

In [7]:
# Calculate the Range and Mean of the Target Variable:
print(f'MAE: {mae}')

y_true = df['actual_sales']

# Calculate the range and mean of the target variable
target_range = np.max(y_true) - np.min(y_true)
target_mean = np.mean(y_true)

# Relative MAE compared to range
relative_mae_range = mae / target_range
print(f"Relative MAE to range: {relative_mae_range:.4f}")

# Relative MAE compared to mean
relative_mae_mean = mae / target_mean
print(f"Relative MAE to mean: {relative_mae_mean:.4f}")

# Evaluate against a baseline model (mean predictor)
baseline_prediction = np.mean(y_true)
baseline_mae = mean_absolute_error(y_true, np.full_like(y_true, baseline_prediction))
print(f"Baseline MAE: {baseline_mae:.4f}")

# Comparing MAE with baseline
improvement_over_baseline = (baseline_mae - mae) / baseline_mae
print(f"Improvement over baseline: {improvement_over_baseline:.4%}")

MAE: 10.0
Relative MAE to range: 0.0333
Relative MAE to mean: 0.0435
Baseline MAE: 96.0000
Improvement over baseline: 89.5833%


A relative MAE of 0.0333 means that the average error is about 3.33% of the range of your target variable. This is quite low, indicating that the errors are small relative to the overall range of ice cream sales.

A relative MAE of 0.0435 means that the average error is about 4.35% of the mean ice cream sales. This is also relatively low, suggesting that the errors are small compared to the average sales.

A baseline MAE of 96 means that a simple mean predictor (predicting the average sales for all temperatures) has an average error of 96 units. Our model's MAE of 10 is significantly better than this baseline.